### ISSUE TITLE
Easier addition of support for newdatatypes

### ISSUE THREAD

**Issue by JervenBolleman** [w3c/sparql-12#130]
## Why?

Currently, most stores require significant work to add new data-types. e.g. anything beyond the inbuild XSD types requires custom code. This makes it more difficult to use datatypes as a significant determinant of meaning.

This come up as part of the solution to issue #129 

## Proposed solution

A RDF file that lists datatypes and can be read by stores.

```turtle
unit:K rdfs:subClassOf xsd:decimal ;
  rdfs:label "Kelvin"@en ;
  rdfs:comment "SI Unit for temperature" .
```

A file such as this adds datatype definitions and allows the mathematical functions of xsd:decimal be called with this value `(sameTerm("273.1"^^unit:K + "100"^^unit:K, "373.1"^^unit:K)`.

## Considerations for backward compatibility

More data in the wild will be inconvenient to use in SPARQL 1.1. endpoints.

**Comment by kasei** [comment]
Would the use of `subClassOf` imply that this would be considered a derived type of decimal? I think it would be really strange to be able to do things like type-promote between Kelvin and decimal, or subtract an integer value from a Kelvin value.

**Comment by VladimirAlexiev** [comment]
How can I express the relation between Kelvin and Celsius?

Or between meter and cm?

--

LINDT shows an example if declaring a datatype and implementing it in JS (see #129).
The real LINDT is implemented in Java.

**Comment by ericprud** [comment]
Interestingly, I'm not sure °C and °K are more related than °C and °F. 

# supportedUnits
At a minimum, we could add a `sd:supportedUnits` property to the [SPARQL Service Description spec](https://www.w3.org/TR/sparql11-service-description/). That would allow clever clients to tailor their queries to whatever the remote endpoint would support.

There's another dimension: which operators are supported. XPath provides names for e.g. lessThan, so a service description might look like:
``` turtle
[] a sd:Service ;
    sd:endpoint <http://www.example/sparql/> ;
    sd:supportedUnits # RDF representation of https://www.w3.org/TR/sparql11-query/#OperatorMapping
      [ sd:left ucum:m ; sd:function op:numericEquals ; sd:right ucum:ft_i ],
      [ sd:left ucum:ft_i ; sd:function op:numericEquals ; sd:right ucum:m ]
      # ...
    .
```
It's kinda tedious to have to write both of those, but maybe we can't assume symmetry.

# unitConversions
The above wouldn't enable clever servers to do automagic conversion. We could pick some base units, e.g. MKS, and have a linear function to capture the mapping a la:
``` turtle
[] a sd:Service ;
    sd:endpoint <http://www.example/sparql/> ;
    sd:unitConversions u:Length, u:Mass, u:Time
    .
```
and centrally maintain the mappings:
``` turtle
u:Length u:baseUnit ucum:m ;
  u:conversion u:Foot, u:Smoot . # ...
u:Mass uLbaseUnit ucum:kg ;
  u:conversion u:Gram, u:Ton, u:LongTon, u:ShortTon, u:Tonne . # ......

u:Foot a u:conversion ; u:factor 0.3048 ; u:offset 0.0 .
u:Smoot a u:conversion ; u:factor 1.7 ; u:offset 0.0 .
...
u: Fahrenheit a u:conversion ; u:factor .555 ; u:offset -17.77 . # assuming offset follows factor.
```

**Comment by JervenBolleman** [comment]
@VladimirAlexiev  @ericprud Easier support of conversion is easier sharing of custom functions/named queries, for which there are few issues already. I wanted to separate out sub parts of the problem to discuss one facet at a time.

@kasei  I am editing the issue to expand the thought behind rdfs:subClassOf

**Comment by afs** [comment]
Should the title be something like:

"Extend SPARQL Service Description to allow declaration of the supported datatypes"

?

**Comment by JervenBolleman** [comment]
@afs no, that is not what I was going for. I was going for a declarative system to declare what properties and operators a new datatype has. i.e. to extends https://www.w3.org/TR/sparql11-query/#matchArbDT. e.g. declare that a new datatype has greater than operator and how that works (in collaboration with issue #131 ) as well as how it can be cast/converted to a different datatype.

**Comment by VladimirAlexiev** [comment]
@JervenBolleman "declare what properties and operators a new datatype has"
@ericprud "RDF representation of https://www.w3.org/TR/sparql11-query/#OperatorMapping"

I agree these would be very useful features. 

---

@maximelefrancois86 and @Antoine-Zimmermann have proposed:
- https://ci.mines-stetienne.fr/lindt/spec.htm:
  - API in WebIDL
  - behavior constraints
  - two implementations: LINDT (discussed in #129) and JS (see below)
- https://ci.mines-stetienne.fr/lindt/v1/custom_datatypes: example of web artefacts delivered with content negotiation:
  - documentation (HTML)
  - declaration (Turtle): but the link is broken because it redirects
  - implementation (JS): but the link is broken because it redirects

---

The standards (esp OWL2) have a lot on datatypes:
- https://www.w3.org/TR/rdf11-concepts/#section-Datatypes
- https://www.w3.org/TR/swbp-xsch-datatypes/#sec-userDefined
- https://www.w3.org/TR/rdf-schema/#ch_datatype
- https://www.w3.org/TR/owl2-quick-reference/#Built-in_Datatypes_and_Facets
  - Built-in Datatypes, including two that are never used: owl:rational, owl:real
  - facets (I think the related reasoning is called D-Entailment)
- https://www.w3.org/TR/owl2-quick-reference/#New_Features_in_OWL_2
  - https://www.w3.org/TR/2012/REC-owl2-new-features-20121211/#Extended_datatype_capabilities
  - Data Ranges:
    - http://www.w3.org/TR/2012/REC-owl2-new-features-20121211/#Datatype_Definitions
    - http://www.w3.org/TR/2012/REC-owl2-new-features-20121211/#Data_Range_Combinations: complement, intersection and union
    - http://www.w3.org/TR/2012/REC-owl2-new-features-20121211/#F10:_Extra_Datatypes_and_Datatype_Restrictions and facets
    - http://www.w3.org/TR/2012/REC-owl2-new-features-20121211/#F11:_N-ary_Datatypes : hook for n-ary datatype
  - Extra Built-in Datatypes: owl:rational, owl:real, xsd:dateTimeStamp, rdf:PlainLiteral
- https://www.w3.org/TR/2012/NOTE-owl2-dr-linear-20121211/ Data Range Extension: Linear Equations

----

Can we try to flesh out a list of requirements? Eg
- subclassing to XSD datatypes
- hooking to SPARQL Description
- hooking to SPARQL Operator Mapping
- API (WebIDL or other)
- declarative description
- a variety of implementations
- web-fetchable implementation
- datatype (data range) restrictions using facets
- central register of datatypes

**Comment by jmkeil** [comment]
Hi. I like the idea and would like to add a requirement for consideration:

### Requirement: Unambiguous definition of conversion values

In an [evaluation of several unit ontologies](http://www.semantic-web-journal.net/system/files/swj1825.pdf), we identified multiple cases of wrong conversion factors caused by mixing up the direction of factor and offset. This mistakes have been made by the people who defined the property. In a standard with wide application, this is even more critical. In fact, factor and offset allow four possible interpretations:
```
a = b × factor + offset
a = (b + offset) × factor
a × factor + offset = b
(a + offset) × factor = b
```
To take the example by @ericprud:
> ```
> u:Fahrenheit a u:conversion ; u:factor .555 ; u:offset -17.77 . # assuming offset follows factor.
> ```
could be less ambiguously expressed e.g. in the following way:
```
u:Fahrenheit u:oneEquals    .55555556^^u:degreeCelsius ;
             u:zeroAt    −17.77777778^^u:degreeCelsius .
```
or
```
u:Fahrenheit u:oneEquals    .55555556 ;
             u:zeroAt    −17.77777778 ;
             u:of        u:degreeCelsius .
```
or similar.

**Comment by steveraysteveray** [comment]
Please take a look at this link
<https://github.com/qudt/qudt-public-repo/wiki/Support-for-measures-of-absolute-values-and-for-intervals-(differences)#converting-absolute-values>
for how I believe unit conversion with offsets should always be calculated.
Your examples above seem to only have one offset value which I find
confusing.

Steve

**Comment by JervenBolleman** [comment]
Unit's are not the only datatypes to consider.  One  I would find very nice to have is conversion between dna/rna plus forward reverse strands etc. in the biological sphere.

**Comment by ericprud** [comment]
With linear numeric units, we can define `oneEquals` and `zeroAt` per [@jmkeil's proposal](#issuecomment-771741209), which allows a naive, generic engine to handle these with no unit-specific code. We won't achieve that for e.g. your example of converting thymine to uracil or mapping [astronomical coordinates](https://heasarc.gsfc.nasa.gov/Tools/name_or_coordinates_help.html), but I think we can still leverage datatypes and operator mappings to advertise capabilities and perform rudimentary unit analysis.

**Comment by jmkeil** [comment]
> Please take a look at this link <https://github.com/qudt/qudt-public-repo/wiki/Support-for-measures-of-absolute-values-and-for-intervals-(differences)#converting-absolute-values> for how I believe unit conversion with offsets should always be calculated. Your examples above seem to only have one offset value which I find confusing. Steve

Yes, there is a difference between absolute values and intervals. But I don't see which information (conversion offset, conversion multiplier) is missing to do both. (I must confess - the property name "oneEquals" does not fit very well to absolute value conversion.) Of course, it would be good practice to define all units in reference to SI base units and I would expect an implementation to combine at least two conversion definitions for conversion between not directly connected units.

One thing, what is missing, is the information when to apply which conversion (absolute or interval). One solution I could think of, is the definition of two datatypes (e.g. `u:degreeCelsiusAbsolute` and `u:degreeCelsiusInterval`). However, this raises new problems on the definition of basic calculations:
* absolute - absolute = interval
* absolute + absolute = ?
* absolute ± interval = absolute
* interval ± interval = interval

**Comment by steveraysteveray** [comment]
Quoting from the [QUDT wiki](https://github.com/qudt/qudt-public-repo/wiki/Support-for-measures-of-absolute-values-and-for-intervals-(differences)#converting-absolute-values),

"To support this, the qudt:Quantity class has a property qudt:isDeltaQuantity. It is associated with the qudt:Quantity because the quantity describes the context of the measurement. qudt:isDeltaQuantity is a boolean property to record whether the measurement is an absolute value of the Quantity instance, or a delta (or difference) value. Setting isDeltaQuantity to "true" means the measurement is an interval. isDeltaQuantity set to "false" means the measurement is an absolute value. An application can then take the appropriate action, such as in unit conversion, etc.

It should be noted that in these cases, the unit is still the same unit on the same scale. There is nothing special about the unit."

**Comment by jmkeil** [comment]
> Quoting from the [QUDT wiki](https://github.com/qudt/qudt-public-repo/wiki/Support-for-measures-of-absolute-values-and-for-intervals-(differences)#converting-absolute-values),
> 
> "To support this, the qudt:Quantity class has a property qudt:isDeltaQuantity. It is associated with the qudt:Quantity because the quantity describes the context of the measurement. qudt:isDeltaQuantity is a boolean property to record whether the measurement is an absolute value of the Quantity instance, or a delta (or difference) value. Setting isDeltaQuantity to "true" means the measurement is an interval. isDeltaQuantity set to "false" means the measurement is an absolute value. An application can then take the appropriate action, such as in unit conversion, etc.
> 
> It should be noted that in these cases, the unit is still the same unit on the same scale. There is nothing special about the unit."

That works out, if a value is represented using an *individual* (represented by IRI or blank node) of a class and with several properties. To represent them as a *literal* (e.g. `"37"^^u:degreeCelsius`), [it isn't an option to add property statements on them](https://www.w3.org/TR/rdf11-concepts/#section-triples)).

**Comment by VladimirAlexiev** [comment]
@jmkeil @steveraysteveray @ericprud  please make a separate issue for discussing conversions (and take a look at the LINDT issue here that's closely related).

Otherwise your valuable comments will be lost in this issue, which is about a different (though related) topic.

**Comment by VladimirAlexiev** [comment]
SciSPARQL is a specialized implementation that includes matrices and tensors and I think will pose strong requirements on this issue.
Eg see
https://ieeexplore.ieee.org/document/6313648 by Andrej Andrejev and Tore Risch.
Does anyone know how to contact them?
